staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/promise/promise.go

```diff
@@ -0,0 +1,78 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package promise
+
+import (
+	"sync"
+)
+
+// promise implements the WriteOnce interface.
+type promise struct {
+	doneCh  <-chan struct{}
+	doneVal interface{}
+	setCh   chan struct{}
+	onceler sync.Once
+	value   interface{}
+}
+
+var _ WriteOnce = &promise{}
+
+// NewWriteOnce makes a new thread-safe WriteOnce.
+//
+// If `initial` is non-nil then that value is Set at creation time.
+//
+// If a `Get` is waiting soon after `doneCh` becomes selectable (which
+// never happens for the nil channel) then `Set(doneVal)` effectively
+// happens at that time.
+//
+// If `doneCh != nil` then `doneVal != nil`.
+func NewWriteOnce(initial interface{}, doneCh <-chan struct{}, doneVal interface{}) WriteOnce {
+	if doneCh != nil && doneVal == nil {
+		panic("NewWriteOnce given nil doneVal")
```

Comment:
I'm really against panicing in situations that may happen as bugs.
[Same below]

Reply:
I'm generally against any panics in non-test code - it makes it impossible to recover in production if something bad happens.

The only situation I can potentially live with are panics within self-contained code [e.g. I have a function that checks something, and later assumes it's true].
Here we can panic on bad use of the type, which is very easy to make a mistake.

Reply:
Returning an error here is a bit artificial - that's why in my code I was passing it in Get() function it makes it more natural to change Get() to return (interface, error) if the passed values are incorrect.